src/accelerate/data_loader.py

```diff
@@ -80,19 +80,13 @@ def __init__(self, *args, **kwargs):
         self.epoch = 0
 
     def __iter__(self):
-        g = torch.Generator()
-        if self.generator is not None:
-            seed = self.epoch + self.generator.initial_seed()
-        else:
-            seed = self.epoch
-        g.manual_seed(seed)
-        n = len(self.data_source)
-        # Taken 1:1 from torch.utils.data.sampler.RandomSampler.__iter__
-        if self.replacement:
-            for _ in range(self.num_samples // 32):
-                yield from torch.randint(high=n, size=(32,), dtype=torch.int64, generator=g).tolist()
-        else:
-            yield from torch.randperm(n, generator=g).tolist()
+        if self.generator is None:
+            self.generator = torch.Generator()
+        # Allow `self.epoch` to modify the seed of the generator
+        seed = self.epoch + self.generator.initial_seed()
+        self.generator.manual_seed(seed)
+        yield from super().__iter__()
+        self.set_epoch(self.epoch + 1)
```

Comment:
Here is what stemmed the original bug/issue, we needed to call `set_epoch` at the end of `__iter__`